File: VotingSystem/src/voting/Voter.java
package voting;

import java.util.ArrayList;
import java.util.List;

public class Voter {

    private String name;
    private String password;
    private long id;
    private static List<Vote> votes = new ArrayList<>();



    public Voter(String name, String password, long voterId) {
        this.name = name;
        this.password = password;
        this.id = voterId;
    }
    public Voter(){

    }


    public long getVoterId() {
        return id;
    }

    public static List<Vote> getVotes() {
        return votes;
    }




    public Vote castVote(long voterId, int candidateId){

        validateVoter(voterId);
        validateCandidate(candidateId);
        validateSpecificCandidateVote(voterId, candidateId);

        Vote vote = new Vote(voterId, candidateId);
        votes.add(vote);

        return vote;
    }

    public void updatePassword(String oldPassword, String newPassword) {
        if(oldPassword.equals(password))this.password = newPassword;
        else throw new IllegalArgumentException("Invalid credentials");
    }



    private boolean validateVoter(long voterId) {
        if(ElectoralBody.getVoterIdList().contains(voterId)){
            return true;
        }

        throw new VoterNotRegisteredException("Voter with id " + voterId + " is not registered");
    }


    private boolean hasVotedASpecificCandidate(long voterId, int candidateId) {
        for (Vote vote : votes) {
            if(vote.getCandidateId() == candidateId && vote.getVoterId() == voterId){
                return true;
            }
        }
        return false;
    }

    private void validateSpecificCandidateVote(long voterId, int candidateId) {
        if(hasVotedASpecificCandidate(voterId, candidateId)){
            throw new CandidateHasBeenVotedException("You have voted already voted for this candidate");
        }
    }

    private void validateCandidate(int candidateId) {
        if(ElectoralBody.getCandidateIdList().contains(candidateId)){
            return;
        }
        throw new CandidateNotRegisteredException("Candidate with id " + candidateId + " is not registered");
    }

}

File: votingsystempython/src/voter.py
from vote import Vote
from electoralbody import ElectoralBody


class Voter:
    def __init__(self, name, id, password, electoral_body)->None:
        self.__name = name
        self.__id = id
        self.__password = password
        self.votes = []
        self.electoral_body = electoral_body


    def get_id(self)->int:
        return self.__id

    def update_password(self, old_password, new_password):
        if old_password == self.__password:
            self.__password = new_password
            return self.__password
        else:
            raise Exception("Password doesn't match")

    def cast_vote(self, candidate_id, voter_id):
        self.validate_voter(voter_id)
        self.validate_candidate_id(candidate_id)
        self.has_voted_a_specific_candidate(candidate_id, voter_id)

        vote = Vote(candidate_id, voter_id)
        self.votes.append(vote)
        return vote

    def has_voted_a_specific_candidate(self, candidate_id, voter_id):
        for vote in self.votes:
            if vote.candidate_id == candidate_id and vote.voter_id == voter_id:
                raise Exception("Can't vote twice")

    def validate_candidate_id(self, candidate_id):
        if int(candidate_id) not in self.electoral_body.get_candidate_id_list():
            raise Exception("Candidate id is not valid")

    def validate_voter(self, voter_id):
        if voter_id not in self.electoral_body.get_voter_id_list():
            raise Exception("Voter id is not valid")











